File: src/utils/jsx.ts
/**
 * Safe interpolation helpers for ExtendScript template literals.
 *
 * ExtendScript snippets are built by interpolating values from tool args into
 * template strings. Without escaping, a layer name containing a single quote
 * or backslash breaks the script (or worse, lets a malicious value inject
 * arbitrary JSX into the running Photoshop). These helpers route every value
 * through `JSON.stringify`, which produces an ExtendScript-parseable string
 * literal with all special characters escaped.
 *
 * Use `jsLit` for any string position and `jsNum` for any numeric position.
 * Never interpolate raw user input into JSX with `${x}` again.
 */

/**
 * Matches any character outside printable ASCII (space through tilde).
 *
 * Module-scoped and /g, so it carries `lastIndex` state. Only ever passed to
 * `String.prototype.replace`, which resets that state; do NOT call `.test()`
 * on it — that would alternate true/false across calls.
 */
const NON_ASCII = /[^ -~]/g;

/**
 * Render a value as a JSX-safe string literal.
 *
 *   const layerName = 'foo"bar';
 *   const jsx = `layer.name = ${jsLit(layerName)};`;
 *   // → layer.name = "foo\"bar";
 */
export function jsLit(value: unknown): string {
  // Every character outside printable ASCII is re-escaped as \uXXXX. The
  // emitted .jsx is written UTF-8 with no BOM and carries no `#encoding`
  // directive, so ExtendScript decodes it by the platform codepage: a raw 'ü'
  // arrives mojibake and every comparison against it misses. The escape is
  // pure ASCII, survives any codepage, and the JS parser turns it back into
  // the exact character. Astral characters need no special case here — JS
  // strings are UTF-16, so each surrogate half escapes separately and the
  // parser rejoins them; the Go twin, which iterates runes, does need one.
  // JSON.stringify has already escaped the control range, so this only ever
  // rewrites characters it left raw — it cannot double-process an escape.
  //
  // Scope: this covers INTERPOLATED VALUES. A snippet's own source text can
  // still carry raw non-ASCII, which this does not touch.
  //
  // Keep in lockstep with jsLit in go-core/jsx.go — the Go core emits the same
  // literals on its own path and a divergence is invisible until a non-ASCII
  // name reaches one emitter but not the other. The two agree on all valid
  // input; they differ only on malformed UTF-8 / lone surrogates, where Go's
  // encoder substitutes U+FFFD and this one preserves the surrogate.
  // tests/unit/jsx.test.ts and go-core/jsx_test.go carry the same case table.
  return JSON.stringify(String(value)).replace(
    NON_ASCII,
    (ch) => `\\u${ch.charCodeAt(0).toString(16).padStart(4, '0')}`
  );
}

/**
 * Render a value as a JSX-safe number literal.
 *
 * If the value isn't a finite number, the fallback is used. This protects
 * against `args.x as number` returning `NaN` and corrupting downstream JSX.
 *
 *   `layer.opacity = ${jsNum(args.opacity, 100)};`
 */
export function jsNum(value: unknown, fallback: number): string {
  const n = typeof value === 'number' ? value : Number(value);
  return Number.isFinite(n) ? String(n) : String(fallback);
}

/**
 * Render a value as a JSX-safe boolean literal (`true` or `false`).
 * Coerces the string forms `'true'` / `'false'` to match how `jsNum`
 * coerces numeric strings — this keeps the helpers symmetric so a value
 * that survived `validateArgs` (which itself accepts both forms) lands
 * here as a real boolean.
 *
 *   `var locked = ${jsBool(args.locked, false)};`
 */
export function jsBool(value: unknown, fallback: boolean): string {
  if (typeof value === 'boolean') return String(value);
  if (value === 'true') return 'true';
  if (value === 'false') return 'false';
  return String(fallback);
}

File: go-core/jsx.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"math"
	"strconv"
	"strings"
)

// Safe interpolation helpers — the Go port of src/utils/jsx.ts. Every value
// that lands in emitted JSX routes through one of these so a layer name with
// a quote/backslash can't break (or inject into) the running script.

// jsLit renders a value as a JSX-safe string literal, mirroring
// `JSON.stringify(String(value))`.
//
// HTML escaping is disabled (SetEscapeHTML(false)) so <, >, & pass through
// literally — matching JS JSON.stringify exactly. Go's encoding/json escapes
// those by default; without this a layer name like "A & B" would round-trip as
// "A & B" and diverge from the TS output for the same input.
func jsLit(s string) string {
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	// Encode of a Go string can never fail: json.Marshal only errors on
	// unsupported types (chan/func/complex), NaN/Inf floats, or cyclic
	// structures — none reachable from a plain string. The error is swallowed
	// deliberately; the empty-buffer fallback below guards against any future
	// surprise so a caller never gets an unquoted (break-out-able) value.
	_ = enc.Encode(s)
	// Encoder.Encode appends a trailing newline; strip it.
	out := strings.TrimRight(buf.String(), "\n")
	if out == "" {
		return `""`
	}
	return asciiEscape(out)
}

// asciiEscape rewrites every character outside printable ASCII as a \uXXXX
// escape.
//
// The emitted .jsx is written UTF-8 with no BOM and carries no `#encoding`
// directive, so ExtendScript decodes it by the platform codepage. A raw 'ü'
// therefore arrives mojibake and every comparison against it misses — the
// inbound half of the same transport problem __notFoundMessage solves on the
// way out. A \uXXXX escape is pure ASCII, survives any codepage, and the JS
// parser turns it back into the exact character.
//
// encoding/json escapes the control range but passes everything above ASCII
// through as raw UTF-8, so this only ever rewrites what it left alone — it
// cannot double-process an escape the encoder already wrote.
//
// Scope: this covers INTERPOLATED VALUES. A snippet's own source text can
// still carry raw non-ASCII, which this does not touch.
//
// Keep in lockstep with jsLit in src/utils/jsx.ts — both emit literals into
// the same scripts on different paths, and a divergence stays invisible until
// a non-ASCII name reaches one emitter and not the other. The two agree on all
// valid input; they differ only on malformed UTF-8 / lone surrogates, where
// encoding/json substitutes U+FFFD and the TS twin preserves the surrogate.
// jsx_test.go and tests/unit/jsx.test.ts carry the same case table.
func asciiEscape(s string) string {
	if isPrintableASCII(s) {
		return s
	}
	var b strings.Builder
	b.Grow(len(s))
	for _, r := range s {
		switch {
		case r >= ' ' && r <= '~':
			b.WriteRune(r)
		case r > 0xFFFF:
			// Outside the BMP: JS string literals address UTF-16 code units,
			// so an astral rune is written as its surrogate pair and the
			// parser rejoins it. Matches what JSON.stringify emits in TS.
			v := r - 0x10000
			fmt.Fprintf(&b, "\\u%04x\\u%04x", 0xD800+(v>>10), 0xDC00+(v&0x3FF))
		default:
			fmt.Fprintf(&b, "\\u%04x", r)
		}
	}
	return b.String()
}

// isPrintableASCII reports whether s is entirely printable ASCII, letting the
// common case skip the rewrite and its allocation.
func isPrintableASCII(s string) bool {
	for i := 0; i < len(s); i++ {
		if s[i] < ' ' || s[i] > '~' {
			return false
		}
	}
	return true
}

// jsNum renders a finite number the way JS String(n) would (shortest
// round-trip). FormatFloat with 'g'/-1 matches JS for the value ranges these
// snippets use (small radii, counts, percentages).
//
// Non-finite guard: strconv.FormatFloat emits NaN/+Inf/-Inf as the bare words
// "NaN"/"+Inf"/"-Inf", which are NOT valid JS numeric literals — interpolated
// into a snippet they'd be parsed as identifiers and throw at runtime (or, for
// "Infinity" lookalikes, silently misbehave). Emitters return a plain string
// with no error channel, so rather than thread an error through dozens of
// callers we coerce a non-finite input to the safe finite literal "0". A NaN
// only reaches here from a malformed caller (JSON numbers are always finite);
// "0" keeps the emitted JSX valid instead of letting an unquoted token break it.
func jsNum(v float64) string {
	if math.IsNaN(v) || math.IsInf(v, 0) {
		return "0"
	}
	return strconv.FormatFloat(v, 'g', -1, 64)
}

// jsInt renders a JSX-safe integer literal. Kept separate from jsNum so an
// index or count can never pick up float formatting ("1e+06") on the way into a
// snippet; ints carry no non-finite case, so no coercion is needed.
func jsInt(v int) string {
	return strconv.Itoa(v)
}

// jsBool renders a JSX-safe boolean literal. The TS jsBool also coerced the
// string forms 'true'/'false'; here params arrive already typed as bool, so a
// direct render matches that path.
func jsBool(v bool) string {
	if v {
		return "true"
	}
	return "false"
}
